File: packages/server/internal/dtos/category.go
package dtos

type Color struct {
	Background string `json:"background"`
	Foregound  string `json:"foreground"`
}

type Category struct {
	Id       int    `json:"id"`
	Label    string `json:"label"`
	Color    Color  `json:"color"`
	IconName string `json:"iconName"`
}

File: packages/web/src/stores/category.ts
import { request } from "@/lib/request";
import { queryOptions } from "@tanstack/react-query";
import z from "zod";

const CATEGORIES = "categories";

const CategorySchema = z.object({
  id: z.number(),
  label: z.string(),
  color: z.object({
    foreground: z.string(),
    background: z.string(),
  }),
  iconName: z.string(),
  order: z.number(),
});

export type Category = z.infer<typeof CategorySchema>;

const fetchAllCategories = () =>
  request("/categories")
    .get(z.array(CategorySchema))
    .then((categories) => categories ?? []);

export const categoriesQuery = queryOptions({
  queryKey: [CATEGORIES],
  queryFn: fetchAllCategories,
});
